lib/core/src/Cardano/Wallet.hs

```diff
@@ -507,9 +519,48 @@ newWalletLayer tracer bp db nw tl = do
         fmap (\(PrimaryKey wid) -> wid) <$> DB.listWallets db
 
     _listTransactions
-        :: WalletId
-        -> ExceptT ErrNoSuchWallet IO [(Tx t, TxMeta)]
-    _listTransactions wid = liftIO $ map snd <$> DB.readTxHistory db (PrimaryKey wid)
+        :: (DefineTx t)
+        => WalletId
+        -> IO [TransactionInfo]
+    _listTransactions wid = assemble <$> DB.readTxHistory db (PrimaryKey wid)
+      where
+        -- This relies on DB.readTxHistory returning all necessary transactions
+        -- to assemble coin selection information for outgoing payments.
+        -- To reliably provide this information, it should be looked up when
+        -- applying blocks, but that is future work (issue #573).
+        assemble :: [(Hash "Tx", (Tx t, TxMeta))] -> [TransactionInfo]
+        assemble txs = map mkTxInfo txs
+          where
+            mkTxInfo (txid, (tx, meta)) = TransactionInfo
+                { txInfoId = txid
+                , txInfoInputs = [(txIn, lookupOutput txIn) | txIn <- W.inputs @t tx]
+                , txInfoOutputs = W.outputs @t tx
+                , txInfoMeta = meta
+                , txInfoDepth = Quantity 0
+                -- fixme: this depth is calculated in slots, but a block depth is required.
+                -- That will require tracking the block height and recording it in TxMeta.
+                , txInfoTime = blockTime (meta ^. #slotId)
+                }
+            txOuts = Map.fromList
+                [ (txid, W.outputs @t tx)
+                | (txid, (tx, _)) <- txs ]
+            -- Because we only track UTxO of this wallet, we can only return
+            -- this information for outgoing payments.
+            lookupOutput (TxIn txid index) =
+                Map.lookup txid txOuts >>= atIndex (fromIntegral index)
+            atIndex i xs = if i < length xs then Just (xs !! i) else Nothing
+
+    -- Calculates the time a block was created. Block creation time is at the
+    -- /end/ of the slot (maybe -- check this).
+    -- There is some silly stuff going on converting 'DiffTime' to
+    -- 'NominalDiffTime'. We should probably use 'UTCTime'/'NominalDiffTime'
+    -- everywhere.
+    blockTime :: SlotId -> UTCTime
+    blockTime sl = addUTCTime (silly offset) block0Date
+      where
+        offset = slotLength * fromIntegral (flatSlot slotsPerEpoch sl + 1)
+        silly :: DiffTime -> NominalDiffTime
+        silly = fromRational . toRational
```

Comment:
Wouldn't it be easier to change the `slotLength` type to be `NominalDiffTime` from the start... :thinking: ? (I think that's what you're suggesting in the comment above, but I am not 100% sure. If so, then I 100% agree and I'd suggest doing it in this PR).

Reply:
Done: 08e5d36f

Reply:
Yes I suggested we should probably use 'UTCTime'/'NominalDiffTime' everywhere, rather than 'UniversalTime'/'DiffTime'.
I assume that, if there is a leap second on December 31, 2019, the slot when midnight ticks over will still be 20 seconds long.